### DISCUSSION TITLE
Price_commodity contains EPS values

### DISCUSSION THREAD

**Discussion by Aprillmylmy** [iiasa/message_ix#644] [MESSAGE modeling] [answered]
Dear developers,

I developed an energy model in messageix, but in _Price_commodity_, EPS is given for selective useful energy commodities. Therefore, Macro will not be able to run through.

Any suggestions on what to check or debug if having EPS in _Price_commodity_? 

Best,
Emily

**Comment by behnam-zakeri** [comment, marked as answer]
@Aprillmylmy thank you for asking this. `PRICE_COMMODITY` in MESSAGEix is calculated based on the marginals of the solution, meaning that it shows the additional cost of the system if one more unit of a certain commodity would be supplied. Therefore, a zero price means that supplying one more unit of that commodity has no additional cost, or simply that commodity is free. If not intended by the user, the zero price of a commodity can have different reasons, including but not limited to:
- the supply source of that commodity is free (e.g., free input commodity for a technology, or free technology itself) ==> Solution: you need to check the cost of the supply technology (or you can check `levelized_cost` from the output GDX).
- there is a sudden reduction in demand so that the supply side cannot adjust, resulting in oversupply. As you may know, the MESSAGEix formulation is based on `supply >= demand`. So, supply can be higher than demand. In some cases, if demand is reduced in a period, and some supply-side technologies cannot reduce their output, the oversupply happens, and the commodity is supplied with zero prices. This can happen when transiting from historical years to the first model year as well, if the values for the `historical_activity` of some technologies are calibrated too high compared to `demand` in the first model period. ==> Solution: check at what level and for what commodity this zero price happens. Then, you can use the MESSAGEix set `balance_equality` to enforce the model to equate supply and demand for that commodity-level. Of course, this may not be the optimal solution (minium cost) but that is another topic beyond this question here.
- Unnecessary bounds on the activity of some technologies: sometimes there are bounds (such as `bound_activity_lo`) that force some technologies to operate even when there is no demand. This can lead to oversupply as well. ==> Solution: check parameters like `bound_*_lo` and make sure the constraints are not too high.
I hope this helps you find the reason for the problem.
